### PR TITLE
VACMS-14442: Enable other CLP sections and test them

### LINKED ISSUE #14442: BE: Create test coverage for CLPs (Campaign Landing Pages)
### User story

**AS A** PO/Editor
**I WANT** the CLP content type covered by automated tests
**SO THAT** any regressions caused by updated code are caught immediately.

### Engineering notes / background

In the course of D10 testing Daniel found that the CLP (Campaign Landing Pages) content type is both complicated and relatively untested.

After collaborating with @ndouglas we decided that we shouldn't add anything to `i_create_a_node.js` for this content type and what is already there was intended for bare minimum, required field testing in lieu of eventually getting to this work we are doing in this ticket.

Adding to `i_create_a_node.js` is fragile and messy and it's worth breaking down what is already in there eventually into other tests.

We should divide up the CLP tests into small, individual tests in `campaign_landing_page.feature`, for differing sections of the CLP page. 

Creating new step definitions that others can also use in `step_definitions/common` will contribute to future forward work instead of adding on top of potential tech debt.

This approach creates a framework for further testing or modifying the campaign landing page in a more efficient and less cognitive way.

## Acceptance criteria

- [x] Researched and implemented proof-of-concept code to write tests in the framework discussed above, which is less flaky and enables reusability
- [ ] Implemented tests in that framework which cover that expected fields are present in the different sections and that enabling the sections works:
   - [ ] Test that expected form elements are present in Hero Banner segment
   - [ ]  Test that expected form elements are present in Why This Matters segment
   - [ ] Test that expected form elements are present in What You Can Do segment
   - [ ] Enable Video page segment and ensure expected fields are present
   - [ ]  Enable Spotlight page segment and ensure expected fields are present
   - [ ]  Enable Stories page segment and ensure expected fields are present
   - [ ]  Enable Downloadable Resources page segment and ensure expected fields are present
   - [ ] Enable Events page segment and ensure expected fields are present
   - [ ] Enable FAQ page segment and ensure expected fields are present
   - [ ] Check that Connect With Us page segment has expected fields/selections

### PR DESCRIPTION
**Pull request by chri5tia**
## Description

Closes #14442 

## Testing done
Local cypress runs

## Screenshots
n/a

## QA steps

Tugboat Cypress tests pass.
Evaluated for false positives.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: tests/cypress/integration/features/content_type/campaign_landing_page.feature
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page.feature:1]: I recommend breaking it up into multiple tests, each beginning with the "I am logged in as a user... I am at /node/add/campaign_landing_page..." bit.

It does add a little bit of overhead to create a user and log in. 

But it also adds better traceability/debugging/maintainability:

- A user is added for each test, so you can see relevant watchdog log messages for a specific user.
- It gives more specificity about where exactly the error occurred -- the test name has more meaning.
- A single Given-When-Then sequence is easier to think about, easier to write about, and easier to map to acceptance criteria (and vice-versa).
- Shorter tests are easier to develop and test independently, since you don't have to wait for _n - 1_ steps to complete before you can see the results of your current area of focus 🙂 

If you have a substantial number of tests, then they can be split into e.g. `campaign_landing_page/hero_banner.feature`, `campaign_landing_page/why_this_matters.feature`, etc. Each of those tests will incur some overhead, but since our tests are parallelized (and ideally will be more parallelized in the future), the length of the entire test run may not increase very much because the tests can be scheduled more efficiently.

File: tests/cypress/integration/features/content_type/campaign_landing_page.feature [hunk 1]
```diff
@@ -1,6 +0,0 @@
-@content_type__campaign_landing_page
-Feature: Content Type: Campaign Landing Page
-
-  Scenario: Log in and create a campaign_landing_page.
-    Given I am logged in as a user with the "content_admin" role
-    Then I create a "campaign_landing_page" node
```
Comment (chri5tia) [inline tests/cypress/integration/features/content_type/campaign_landing_page.feature:6]: This gets moved to `/campaign_landing_page/basic_requirements.feature`

File: tests/cypress/integration/step_definitions/common/i_enable_the_segment.js
Comment (ndouglas) [inline tests/cypress/integration/step_definitions/common/i_enable_the_segment.js:1]: **Praise**: Nice!

File: tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+Feature: Content Type: Campaign Landing Page
+
+  Scenario: Test that expected form elements are present in Why This Matters segment
+    Given I am logged in as a user with the "content_admin" role
+    And I am at "node/add/campaign_landing_page"
+    And I click to expand "Why this matters"
+    Then I fill in field with selector "#edit-field-clp-why-this-matters-0-value" with fake text
+    And I click to expand "Select up to 3 audiences"
+    Then I click the "Select audiences" button
+    And I wait "3" seconds
```
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature:11]: **Suggestion**: This would be a great place for an 'I wait for "{selector}" to [dis]appear' step definition.

File: tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    And I make a selection
+    And I click the "Select audiences" button
+    And I wait "3" seconds
```
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature:14]: **Suggestion**: These fixed waits will make the tests flaky. We should wait on <something> to happen.

**Review by ndouglas — CHANGES_REQUESTED** [review]
This looks fantastic overall.

My main suggestion is to wait for elements to appear or disappear, etc, rather than using wait times. The wait times will make the tests flaky.

File: tests/cypress/integration/features/content_type/campaign_landing_page/basic_requirements.feature
Comment (dsasser) [inline tests/cypress/integration/features/content_type/campaign_landing_page/basic_requirements.feature:1]: Question:
Do we plan to remove the redundant bits from i_create_a_node.js that are now covered by the new tests?

Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/basic_requirements.feature:1]: The new tests do not submit the form, they just check for the presence and function of expected elements. I'm still figuring out if there is anything in the the old test that can be removed with that in mind. Maybe in a future ticket.

Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/basic_requirements.feature:1]: But in short, yes!

File: tests/cypress/integration/step_definitions/common/i_check_the_checkbox.js [hunk 4]
```diff
@@ -27,3 +27,7 @@ Given(`I check the checkbox with selector {string}`, (selector) => {
 Given(`I uncheck the checkbox with selector {string}`, (selector) => {
   cy.get(selector).uncheck({ force: true });
 });
+
+Given("I make a selection", () => {
+  cy.get('input[type="checkbox"]').eq(2).check({ force: true });
```
[the remark this answers is not in the document]
Comment (chri5tia) [reply tests/cypress/integration/step_definitions/common/i_check_the_checkbox.js:32]: The idea behind this is to just select a "random" option. It's great for lists generated by taxonomy that is liable to change without needing to affect the test. It should be selecting the second option, which also ensures there are actual options.

File: tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature (hunk 2 above)
Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature:11]: I love this. I attempted to use `cy.waitUntil` but apparently it's not installed? Unless I used it incorrectly, it's saying that waitUntil isn't a function:
https://www.npmjs.com/package/cypress-wait-until

Here is an example:

```
Given(
  "I wait until the element with the selector {string} is visible",
  (selector) => {
  cy.waitUntil(selector).should("be.visible");
});
```

Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature]: Done. However, I am realizing that there were some problems with how I was testing the modal e.g. false positives. I have removed all the modals from the test for now until I can do more work to see how to get cypress and iframes to play nice.

File: tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature (hunk 3 above)
Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/why_this_matters.feature]: These are removed in lieu of removing modal tests.

File: tests/cypress/integration/features/content_type/campaign_landing_page/faq.feature [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    And I click to collapse "Hero banner"
+    And I click to expand "FAQs"
+    And I enable the page segment
+    Then I click the "Add Q&A" button
+    And I fill in "Question" field with fake text
+    And I fill in "Text" field with fake text
+    And I should see "Add a link to more FAQs"
+#    Then I click the "Add Call to action" button
+#    And I fill in "Link" field with fake link
+#    And I fill in "Link text" field with fake text
```
[the remark this answers is not in the document]
Comment (chri5tia) [reply tests/cypress/integration/features/content_type/campaign_landing_page/faq.feature]: I was having trouble getting cypress to find these fields.

File: tests/cypress/integration/step_definitions/common/i_check_the_checkbox.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (chri5tia) [reply tests/cypress/integration/step_definitions/common/i_check_the_checkbox.js]: Turns out it doesn't work. I don't know if it's because I switched to Electron instead of Chrome but I suspect it needed to be tested more so I removed it.

**Comment by chri5tia** [comment]
@ndouglas Can you take a fresh look? There are two features that need more work for accurate testing: CTA buttons and Add media button/iframe modals. I have put TODO items in place so that we can keep track of what is missing. Let me know if I should remove those before merging and if anything else is needed. I think this is a logical stopping point for this ticket. Let me know your thoughts.

**Comment by chri5tia** [comment]
Creating new tickets for testing the modal and testing the CTA links

File: tests/cypress/integration/features/content_type/campaign_landing_page/connect_with_us.feature
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/connect_with_us.feature:1]: **Suggestion**: Behavioral tests normally follow a Given-When-Then format, and this follows a sort of Given-Then format.

Given should describe preconditions, initial setup, etc.

When should describe the actions that you take.

Then should describe the expectations of the effects of those actions.

So in this case, I'd say "Given I am logged in... and I am at... and I click to expand..."

The "When" is fuzzier. Here, it just seems like you're demonstrating that a dropdown works. I'd suggest formatting it like this:

- Given... as you have it.
- When I click to expand "Connect with us"
- Then the dropdown with selector "#edit-field-related-office" has the option "NCA"
- And the dropdown with selector "#edit-field-related-office" has the option "VAC0"
- etc.

That's then a little clearer about what you're wanting to see here.

File: tests/cypress/integration/features/content_type/campaign_landing_page/spotlight.feature
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/spotlight.feature:1]: **Suggestion**: This is a Given-Then-Then test. 

If you have a Then section, you can add another Then section... but you should add another When section before it to specify what exactly you're doing.

e.g.

- Given I am standing in my yard
- When I look up
- Then I should see the sky
- When I look down
- Then I should see the ground
- When I look in a mirror
- Then I should see an idiot
- etc

I think it should look like:

- Given I am logged in...
- When I fill in field...
- Then field with selector "#whatever" contains some text
- When I click the "Add Link teaser" button...
- And I ...
- Then "Link summary" should contain some text

Something like that.

**Review by ndouglas — DISMISSED** [review]
This looks really good, but it's much better if you follow a Given-When-Then format rather than Given-Then or Given-Then-Then. Please change those where possible.

In some cases, I ask you to write step definitions to verify that text has been written successfully to a field (rather than just disappearing into the DOM or the void or something). That's not necessary right now, but would be a great place to follow up, just so there's some actual verification in a Then step happening.

So, bottom line is I'd like you to make some changes with how you use Given-When-Then, but that should mostly just be some light editing and shouldn't involve making any substantial changes, adding additional tests or step definitions, etc. Feel free to open up followup issues for those, or don't.

Thank you!

**Comment by chri5tia** [comment]
@ndouglas Thanks for explaining the Given, When, Then flow. I fixed them for each of the tests and it will hopefully be more apparent that I am just testing if I can fill out a field or if a field is there. At this point, nothing is being submitted because I haven't figured out how to resolve all the redundancy for filling out the required fields and also I think the modal and CTA tests would need to be in place first. But right now in some places seeing if it can fill out or interact with a field is like seeing if it's there, its the kind of field that was expected and permissions are correct, etc. Can you take a fresh look with these changes?

File: tests/cypress/integration/features/content_type/campaign_landing_page/connect_with_us.feature
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/connect_with_us.feature:1]: **Praise**: Yep, that's great!

File: tests/cypress/integration/features/content_type/campaign_landing_page/downloadable_resources.feature [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    Then I should see an element with the selector "#edit-field-clp-resources-header-0-value"
+    And I can fill in field with selector "#edit-field-clp-resources-header-0-value" with fake text
+    And I can fill in field with selector "#edit-field-clp-resources-intro-text-0-value" with fake text
+    When I click the "Add new Downloadable resource" button
+    Then I can fill in "Name" field with fake text
+    And I can fill in "External File URL" field with fake text
+    And I can fill in "Description" field with fake text
+# TODO: Test section drop-down for this segment
+    And I click the "Cancel" button
+    Then I click the "Add existing Downloadable resource" button
```
Comment (ndouglas) [inline tests/cypress/integration/features/content_type/campaign_landing_page/downloadable_resources.feature:19]: **Nitpick**: Missed one.

**Review by ndouglas — APPROVED** [review]
These look great! Thank you!

**Comment by ndouglas** [comment]
Looks like there's an issue with the video test:

https://cms-bk1bdtbdtzgaipprqu6gzyeortznilju.ci.cms.va.gov/cypress/videos/content_type/campaign_landing_page/video.feature.mp4
https://cms-bk1bdtbdtzgaipprqu6gzyeortznilju.ci.cms.va.gov/cypress/screenshots/actual/content_type/campaign_landing_page/video.feature/Content%20Type%20Campaign%20Landing%20Page%20--%20Enable%20Video%20page%20segment%20and%20ensure%20expected%20fields%20are%20present%20(failed).png

**Comment by chri5tia** [comment]
See this PR to fix video etc: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/15028
